pennylane/devices/jax_ops.py

```diff
@@ -0,0 +1,169 @@
+
+
+def RY(theta):
+    r"""One-qubit rotation about the y axis.
+
+    Args:
+        theta (float): rotation angle
+
+    Returns:
+        array[complex]: unitary 2x2 rotation matrix :math:`e^{-i \sigma_y \theta/2}`
+    """
+    return jnp.cos(theta / 2) * I + 1j * jnp.sin(-theta / 2) * Y
+
+
+def RZ(theta):
+    r"""One-qubit rotation about the z axis.
+
+    Args:
+        theta (float): rotation angle
+
+    Returns:
+        array[complex]: the diagonal part of the rotation matrix :math:`e^{-i \sigma_z \theta/2}`
+    """
+    p = jnp.exp(-0.5j * theta)
+    return jnp.array([p, jnp.conj(p)])
+
+
+def Rot(a, b, c):
+    r"""Arbitrary one-qubit rotation using three Euler angles.
+
+    Args:
+        a,b,c (float): rotation angles
+
+    Returns:
+        array[complex]: unitary 2x2 rotation matrix ``rz(c) @ ry(b) @ rz(a)``
+    """
+    return jnp.diag(RZ(c)) @ RY(b) @ jnp.diag(RZ(a))
+
+
+def CRX(theta):
+    r"""Two-qubit controlled rotation about the x axis.
+
+    Args:
+        theta (float): rotation angle
+
+    Returns:
+        array[complex]: unitary 4x4 rotation matrix
+        :math:`|0\rangle\langle 0|\otimes \mathbb{I}+|1\rangle\langle 1|\otimes R_x(\theta)`
+    """
+    return (
+        jnp.cos(theta / 4) ** 2 * II
+        - 1j * jnp.sin(theta / 2) / 2 * IX
+        + jnp.sin(theta / 4) ** 2 * ZI
+        + 1j * jnp.sin(theta / 2) / 2 * ZX
+    )
+
+
+def CRY(theta):
+    r"""Two-qubit controlled rotation about the y axis.
+
+    Args:
+        theta (float): rotation angle
+    Returns:
+        array[complex]: unitary 4x4 rotation matrix :math:`|0\rangle\langle 0|\otimes \mathbb{I}+|1\rangle\langle 1|\otimes R_y(\theta)`
+    """
+    return (
+        jnp.cos(theta / 4) ** 2 * II
+        - 1j * jnp.sin(theta / 2) / 2 * IY
+        + jnp.sin(theta / 4) ** 2 * ZI
+        + 1j * jnp.sin(theta / 2) / 2 * ZY
+    )
+
+
+def CRZ(theta):
+    r"""Two-qubit controlled rotation about the z axis.
+
+    Args:
+        theta (float): rotation angle
+    Returns:
+        array[complex]: diagonal part of the 4x4 rotation matrix
+        :math:`|0\rangle\langle 0|\otimes \mathbb{I}+|1\rangle\langle 1|\otimes R_z(\theta)`
+    """
+    p = jnp.exp(-0.5j * theta)
+    return jnp.array([1.0, 1.0, p, jnp.conj(p)])
+
+
+def CRot(a, b, c):
+    r"""Arbitrary two-qubit controlled rotation using three Euler angles.
+
+    Args:
+        a,b,c (float): rotation angles
+    Returns:
+        array[complex]: unitary 4x4 rotation matrix
+        :math:`|0\rangle\langle 0|\otimes \mathbb{I}+|1\rangle\langle 1|\otimes R(a,b,c)`
+    """
+    return jnp.diag(CRZ(c)) @ (CRY(b) @ jnp.diag(CRZ(a)))
+
+
+def MultiRZ(theta, n):
+    r"""Arbitrary multi Z rotation.
+
+    Args:
+        theta (float): rotation angle :math:`\theta`
+        wires (Sequence[int] or int): the wires the operation acts on
+    Returns:
+        array[complex]: diagonal part of the multi-qubit rotation matrix
+    """
+    return jnp.exp(-1j * theta / 2 * pauli_eigs(n))

```

Comment:
This file is *so* identical to the Autograd version, I'm tempted to suggest merging it somehow. But that seems fraught/the wrong abstraction, so I'm resisting.

Reply:
I would agree, lets keep them separate. There are I think 3-4 subtle changes, and making autograd and jax share exact codepaths is a little rough. We did something similar with TensorNetwork back in the day (the before times), where our jax backend would inherit the numpy backend, and we ended up breaking them apart later anyway.

Reply:
the before times lol